Allow overshooting final htlc amount and expiry

alecchendev opened the pull request:
Closes #1872, #2012 (implements lightning/bolts#1031, lightning/bolts#1032, lightning/bolts#1040)

The rationale/changes for the spec PRs, respectively:

While retrying a failed path of an MPP, a node may want to overshoot the `total_msat` in order to use a path with an `htlc_minimum_msat` greater than the remaining value being sent. This PR no longer fails MPPs that overshoot the `total_msat`, however it does fail HTLCs with the same payment hash that are received *after* a payment has become claimable.

Final nodes previously had stricter requirements on HTLC contents matching onion values compared to intermediate nodes. This allowed for probing, i.e. the last intermediate node could overshoot the value by a small amount and conclude from the acceptance or rejection of the HTLC whether the next node was the destination. This PR now allows overshooting the `outgoing_cltv_value`. This also applies to the `amount_msat` and `amt_to_forward`, however this change was already present.

If routing nodes take less fees and pay the final node more than `amt_to_forward`, the receiver may see that `total_msat` has been met before all of the sender's intended HTLCs have arrived. The receiver may then prematurely claim the payment and release the payment hash, allowing routing nodes to claim the remaining HTLCs. This PR now uses the onion value `amt_to_forward` to determine when `total_msat` has been met which the sender to control the set total.

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2584,7 +2584,7 @@ where
 
 	#[cfg(test)]
-	fn test_send_payment_internal(&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>, keysend_preimage: Option<PaymentPreimage>, payment_id: PaymentId, recv_value_msat: Option<u64>, onion_session_privs: Vec<[u8; 32]>) -> Result<(), PaymentSendFailure> {
+	pub(super) fn test_send_payment_internal(&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>, keysend_preimage: Option<PaymentPreimage>, payment_id: PaymentId, recv_value_msat: Option<u64>, onion_session_privs: Vec<[u8; 32]>) -> Result<(), PaymentSendFailure> {
```
I could avoid doing this by moving my test into the channel manager tests, but I initially put mine in `functional_tests.rs` because the tests I modeled mine off of were in there. Open to suggestions

alecchendev commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -915,7 +915,6 @@ impl OutboundPayments {
 			return Err(PaymentSendFailure::PathParameterError(path_errs));
 		}
 		if let Some(amt_msat) = recv_value_msat {
-			debug_assert!(amt_msat >= total_value);
```
Hopefully this won't disrupt other testing much, I couldn't find any other way to send an MPP while being able to set the `total_msat` to be lower for testing, and all of the send payment methods seem to call `pay_route_internal` anyway so I figured I might not be able to avoid this.

alecchendev reviewed (commented):
Some notes for anyone reviewing :)

freddiekrugerrand reviewed (commented):
Why specifically only allow overshooting by the smallest HTLC amount? The spec change doesn't say that there should be a limit on overpayment.

Seems like it would be a much simpler change to just accept if `> total_msat`? Also prevents fingerprinting of LDK based nodes (if everyone else accepts overpayment of any amount and LDK doesn't, that identifies it).

alecchendev commented:
> Why specifically only allow overshooting by the smallest HTLC amount? The spec change doesn't say that there should be a limit on overpayment.

I was mainly going off the part in [the rationale](https://github.com/lightning/bolts/blob/master/04-onion-routing.md#rationale) where it says "However, no individual HTLC may be for less than the difference between the total paid and `total_msat`." I figured the receiving node would have some restriction enforcing that, but I could be misinterpreting.

> Seems like it would be a much simpler change to just accept if `> total_msat`? Also prevents fingerprinting of LDK based nodes (if everyone else accepts overpayment of any amount and LDK doesn't, that identifies it).

Yea that's a good point. In hindsight it might've been better to start with the simpler change and ask if the extra restriction should be added. I haven't checked what the other implementations do so I can look into that soon. If other people agree I can remove the extra restriction :)

alecchendev commented:
> Why specifically only allow overshooting by the smallest HTLC amount?

Just thinking more about this, why would nodes want to overshoot by an entire HTLC? If nodes overshoot by that much, I figure they should just make another payment? Since relaxing this requirement is mainly to serve nodes retrying payments using a route with a constraining `htlc_minimum_msat`, I think it makes sense to only allow this, but again I'm open to further discussion

freddiekrugerrand commented:
> Just thinking more about this, why would nodes want to overshoot by an entire HTLC? 

I agree that this is pretty unlikely, and they _should_ just send another payment if they want to over pay. I just don't think it's valuable to _enforce_ that, since they do no harm by giving us more money and the simpler change is (iiuc) still spec-compliant :)

alecchendev commented:
Made the aforementioned changes, removed some tests that no longer seem necessary, and updated the PR description to reflect these changes!

alecchendev commented:
Added in changes for failing HTLCs for already claimable payments, as well as for https://github.com/lightning/bolts/pull/1032 and https://github.com/lightning/bolts/pull/1040. Might've missed something, but several of the changes from the spec PRs seemed to already be present in the code.

In case it helps save anyone some time looking, here's where I found some of the checks:

Checking `amount_msat < amt_to_forward` and `cltv_expiry < outgoing_cltv_value` were both in `ChannelManager::construct_recv_pending_htlc_info`, while `cltv_expiry < current_block_height + min_final_cltv_expiry_delta` seems it can be inferred from a check in that function, plus in `ChannelManager::decode_update_add_htlc_onion` (line 2279) --> `Channel::internal_htlc_satisfies_config`.

For 1040 the HTLC values that are being used to calculate the MPP set total were already coming from the `outgoing_amt_msat`, but I had to change one spot where that value had been initialized as the `amount_msat`.

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3935,7 +3939,7 @@ where
 			return;
 		}
-		if claimable_amt_msat != expected_amt_msat.unwrap() {
+		if claimable_amt_msat < expected_amt_msat.unwrap() {
```
Hm, it does look redundant. I tried taking it out to see and it seems it's relevant in the functional test `test_double_partial_claim`, in the case where HTLCs timeout before the user claims their funds, then the sender resends only part of the payment, and *then* the user tries to process the original `PaymentClaimable` event and claim the payment. I guess were you talking about if the check was necessary or something else?

alecchendev replied:
Oh gotcha thanks for the clarification

alecchendev replied:
Agh my bad, sorry for misunderstanding again, what I thought you were talking about in this comment was that `claimable_amt_msat` had been changed to be the sum of the onion values (as `htlc.value` had been changed to be `outgoing_amt_msat`, i.e. `amt_to_forward` from the onion), so I changed `htlc.value` back to be the received amount (and added `sender_intended_value` for the onion amount to use in MPP set calculation).

IIUC I don't think `total_msat` was changed in this PR, just the validation against it, so I believe `htlc.total_msat`/`expected_amt_msat` was always the value in the onion?

Now that I get what you're saying however, I agree this check would make more sense if `expected_amt_msat` matched the amount reported in `PaymentClaimable`. Happy to add this here, or in a followup.

alecchendev replied:
> `total_msat` is still calculated as `htlc.total_msat` for MPP and `outgoing_amt_msat` for non-MPP, the problem is `outgoing_amt_msat` has changed here - its now the onion data rather the incoming amount. But, more importantly, the event has changed - the event now tells the user about the actual amount,

Okay, *noww* I see. Thank you so much for explaining!

> Let's just do it here?

On it!

alecchendev replied:
Thinking about this again, should I actually just change the amount reported in `PaymentClaimable` back to be the amount from the onion (i.e. `htlc.total_msat` for MPP and `outgoing_amt_msat` for non-MPP)? I'm thinking that will accomplish we want here: when claiming funds, `expected_amt_msat` will represent what was reported to the user in `PaymentClaimable`. The downside with this is that the event reports the amount from the invoice, while the actual amount may overshoot that.

I'm asking this because as I was trying to change `expected_amt_msat` to match `PaymentClaimable` now that `PaymentClaimable` reports the actual amount received, I'm realizing it introduces this case: a node overpays the original amount, `PaymentClaimable` is generated with the actual received amount, one/some of the HTLCs timeout before the receiver handles the event, now the sender resends HTLCs however now these HTLCs pay enough for the invoice but below `expected_amt_msat`, i.e. the amount previously reported in `PaymentClaimable`, so it refuses to claim. I'm thinking it makes more sense to be able to claim as long as we've received at least the amount originally requested in the invoice, and `PaymentClaimable` should just report that and sometimes the user will end up claiming a little extra?

alecchendev replied:
> We could also report the sender-reported amount to the user if we want, but I think that's confusing (its neither the amount they received nor the amount they specified in the invoice...but something else).

Ah, true.

> That's a good point, but I don't think we need to worry too much about this. If it causes payment failure, so be it - its generally not something that should ever happen HTLC timeouts after receiving them should be relatively rare, only when they were super tight to begin with and we get a block quickly thereafter. We could also consider making it arbitrarily rare by adding an extra block or two of headroom when accepting the HTLC vs when we go to time it out (I think I'll probably do that in a followup as a part of https://github.com/lightningdevkit/rust-lightning/issues/2017 anyway).

Okay gotcha, sounds good :)

alecchendev commented:
I appreciate the feedback! Will try to update this soon

alecchendev commented:
Added a test for overshooting final cltv expiry, and went back to having an `assert_eq!` when comparing the `PaymentClaimable` amount to test those values to make sure they're the actually received amount and not the sender's onion amounts.

To make `PaymentClaimable` use the received amount while still validating MPPs on the intended amount in the onion, and allow `claim_funds` to use the received amount, I added a new field to `ClaimableHTLC` - `intended_value`. My intention was for it to be forwards + backwards compatible with a default value of `value` if it wasn't found. Based on how I understand the tlv macros I think what I did should do that(?), wasn't sure if there is a good way to test that within the repo

valentinewallace reviewed (commented):
Nice! Just nits

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3389,13 +3397,14 @@ where
 															receiver_node_id: Some(receiver_node_id),
 															payment_hash,
-															amount_msat: outgoing_amt_msat,
+															amount_msat,
```
Originally had this but `claimable_htlc` gets moved in the insert

alecchendev commented:
Awesome, thanks everyone for the review!

valentinewallace reviewed (dismissed):
Good stuff!!

alecchendev commented:
Just added a commit before everything to be able to validate on the amount reported in `PaymentClaimable` so I wasn't undoing work done in previous commits

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2159,7 +2165,7 @@ where
 			incoming_shared_secret: shared_secret,
 			incoming_amt_msat: Some(amt_msat),
-			outgoing_amt_msat: amt_msat,
+			outgoing_amt_msat: hop_data.amt_to_forward,
```
nit: I think it would be helpful to document on `PendingHTLCInfo` that on receive, `incoming_amt_msat` refers to the actual amount received and `outgoing_amt_msat` refers to the sender-intended amount

freddiekrugerrand reviewed (commented):
I have limited understanding of channelmanager, but lgtm + nicely tested/structured change

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3328,11 +3328,13 @@ where
 											fail_htlc!(claimable_htlc, payment_hash);
-										} else if total_value == $payment_data.total_msat {
+										} else if total_value - claimable_htlc.value >= $payment_data.total_msat {
+											log_trace!(self.logger, "Failing HTLC with payment_hash {} as payment is already claimable",
+												log_bytes!(payment_hash.0));
```
I think this is actually already covered in `functional_tests::test_dup_htlc_second_rejected` :)

alecchendev commented:
Rebased because of conflicts

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3324,18 +3338,22 @@ where
+											log_trace!(self.logger, "Failing HTLC with payment_hash {} as payment is already claimable",
+												log_bytes!(payment_hash.0));
+											fail_htlc!(claimable_htlc, payment_hash);
+										} else if total_value >= $payment_data.total_msat {
```
nit: Think we're just missing the comment here that references back to the timeout (which you've commented already) :)

alecchendev replied:
I added it above the first if statement because I thought it'd be weird with the indenting to include it in inside the branch above - in hindsight I think that actually would've been better, but since it's there at least I think I'll leave it as is :)

dunxen reviewed (approved):
Nice! Feel free to just address that nit in a follow-up if you want. Don't want to hold this back longer just for that :)

valentinewallace commented:
Merged since I want to pull these changes into some of my PRs, but feel free to address @dunxen's comment in a follow-up (or not, since it's a nit :))
